extensions/sql-database-projects/src/models/deploy/deployService.ts

```diff
@@ -0,0 +1,403 @@
+			return connectionUrl;
+		} else {
+			return undefined;
+		}
+	}
+
+	private async validateConnection(connection: ConnectionResult | string | undefined): Promise<ValidationResult> {
+		const getAzdataApi = await utils.getAzdataApi();
+		if (!connection) {
+			return { validated: false, errorMessage: constants.connectionFailedError('No result returned') };
+		} else if (getAzdataApi) {
+			const connectionResult = <ConnectionResult>connection;
+			if (connectionResult) {
+				const connected = connectionResult !== undefined && connectionResult.connected && connectionResult.connectionId !== undefined;
+				return { validated: connected, errorMessage: connected ? '' : constants.connectionFailedError(connectionResult?.errorMessage) };
+			} else {
+				return { validated: false, errorMessage: constants.connectionFailedError('') };
+			}
+		} else {
+			return { validated: connection !== undefined, errorMessage: constants.connectionFailedError('') };
+		}
+	}
+
+	private async formatConnectionResult(connection: ConnectionResult | string | undefined): Promise<string> {
+		const getAzdataApi = await utils.getAzdataApi();
+		const connectionResult = connection !== undefined && getAzdataApi ? <ConnectionResult>connection : undefined;
+		return connectionResult ? connectionResult.connectionId : <string>connection;
+	}
+
+	public async getConnection(profile: ILocalDbSetting, savePassword: boolean, database: string, timeoutInSeconds: number = 5): Promise<string | undefined> {
+		const getAzdataApi = await utils.getAzdataApi();
+		let connection = await this.retry(
+			constants.connectingToSqlServerOnDockerMessage,
+			async () => {
+				return await this.connectToDatabase(profile, savePassword, database);
+			},
+			this.validateConnection,
+			this.formatConnectionResult,
+			5, timeoutInSeconds);
+
+		if (connection) {
+			const connectionResult = <ConnectionResult>connection;
+			if (getAzdataApi) {
+				return await getAzdataApi.connection.getUriForConnection(connectionResult.connectionId);
+			} else {
+				return <string>connection;
+			}
+		}
+
+		return undefined;
+	}
+
+	private async executeTask<T>(taskName: string, task: () => Promise<T>): Promise<T> {
+		const getAzdataApi = await utils.getAzdataApi();
+		if (getAzdataApi) {
+			return new Promise<T>((resolve, reject) => {
+				let msgTaskName = taskName;
+				getAzdataApi!.tasks.startBackgroundOperation({
+					displayName: msgTaskName,
+					description: msgTaskName,
+					isCancelable: false,
+					operation: async op => {
+						try {
+							let result: T = await task();
+
+							op.updateStatus(getAzdataApi!.TaskStatus.Succeeded);
+							resolve(result);
+						} catch (error) {
+							let errorMsg = constants.taskFailedError(taskName, error ? error.message : '');
+							op.updateStatus(getAzdataApi!.TaskStatus.Failed, errorMsg);
+							reject(errorMsg);
+						}
+					}
+				});
+			});
+		} else {
+			return await task();
+		}
+	}
+
+	private async retry<T>(
```

Comment:
This kind of thing should definitely go into a utils helper module - it doesn't need to be a part of any class.